README.md
# Fhferreira/Round

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

Round numbers with normatives like ABNT NBR 5891.

## Structure

```
src/
tests/
```

## Install

Via Composer

``` bash
$ composer require fhferreira/round
```

## Usage

``` php
use Fhferreira\Round\AbntNbr5891;

echo AbntNbr5891::round(12.744623);
// it should echo 12.74
```

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email flaviometalvale@gmail.com instead of using the issue tracker.

## Credits

- [Flávio H. Ferreira][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/fhferreira/round.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/fhferreira/round/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/fhferreira/round.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/fhferreira/round.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/fhferreira/round.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/fhferreira/round
[link-travis]: https://travis-ci.org/fhferreira/round
[link-scrutinizer]: https://scrutinizer-ci.com/g/fhferreira/round/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/fhferreira/round
[link-downloads]: https://packagist.org/packages/fhferreira/round
[link-author]: https://github.com/fhferreira
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `round` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/ExampleTest.php
<?php

declare(strict_types=1);

namespace Fhferreira\Round;

class ExampleTest extends \PHPUnit\Framework\TestCase
{
    /**
     * Test that true does in fact equal true
     */
    public function testTrueIsTrue()
    {
        $this->assertTrue(true);
    }
}


tests/AbntNbr5891Test.php
<?php

declare(strict_types=1);

namespace Fhferreira\Round\Tests;

use Fhferreira\Round\AbntNbr5891;
use Fhferreira\Round\Exceptions\CommaException;

class AbntNbr5891Test extends \PHPUnit\Framework\TestCase
{
    /**
     * A basic unit test example.
     *
     * @return void
     */
    public function testAmounts()
    {
        echo PHP_EOL;

        $amounts = array(
            //value, expected
            [12.6529, 12.65],
            [12.86512, 12.87],
            [12.744623, 12.74],
            [12.8752, 12.88],
            [12.8150, 12.82],
            [12.8050, 12.80],
            [12.5, 12.50],
            [12.3, 12.30],
            [12.33, 12.33],
            [12, 12],
            [0.025, 0.02],
            [0.036, 0.04],
            [0.046, 0.05],
            [0.026, 0.03],
            [0.2345645, 0.23],
        );

        foreach ($amounts as $index => $amountInfo) {
            $rounded = AbntNbr5891::round($amountInfo[0]);
            $this->assertEquals($amountInfo[1], $rounded);
        }
    }

    public function testPercentFromABNT()
    {
        $amounts = array(
            //value, percent, expected
            [100, 12, 12],
            [100, 13.45, 13.45],
            [149.57, 60, 89.74],
            [149.57, 40, 59.83],
            [4.54, 40, 1.82],
            [4.54, 60, 2.72],
            [159.38, 40, 63.75],
            [159.38, 60, 95.63],
            [159.38, 35, 55.78],
            [159.38, 65, 103.60],
        );

        foreach ($amounts as $index => $amountInfo) {
            $rounded = AbntNbr5891::round(AbntNbr5891::getPercentFromAmount($amountInfo[0], $amountInfo[1]));
            $this->assertEquals($amountInfo[2], $rounded);
        }
    }

    public function testPercentFromABNTPlus100()
    {
        $amounts = array(
            //value, percent, expected
            [100, 12, 12, 1200],
            [100, 13.45, 13.45, 1345],
            [149.57, 60, 89.74, 8974],
            [149.57, 40, 59.83, 5983],
            [4.54, 40, 1.82, 182],
            [4.54, 60, 2.72, 272],
            [159.38, 40, 63.75, 6375],
            [159.38, 60, 95.63, 9563],
            [159.38, 35, 55.78, 5578],
            [159.38, 65, 103.60, 10360],
        );

        foreach ($amounts as $index => $amountInfo) {
            $rounded = AbntNbr5891::round(AbntNbr5891::getPercentFromAmount($amountInfo[0], $amountInfo[1]));
            $this->assertEquals($rounded * 100, $amountInfo[3]);
        }
    }

    /**
     * @expectedException CommaException
     */
    public function testCommaException()
    {
        try {
            $rounded = AbntNbr5891::round('100,50');
            $this->fail("Expected CommaException has not been raised.");
        } catch (CommaException $ex) {
            $this->assertEquals($ex->getMessage(), 'Comma-formatted numbers are not allowed.');
        }
    }
}


src/AbntNbr5891.php
<?php

declare(strict_types=1);

namespace Fhferreira\Round;

use Fhferreira\Round\Exceptions\CommaException;

/**
 * Class Rounder
 * @package Fhferreira\Round
 *
 * The rounding rules, following the ABNT NBR 5891 Standard, apply to the decimal digits located in the position
 * following the number of decimal digits to be transformed, that is, if we have a number of 4, 5, 6, n decimal digits
 * and if we want to round to 2, these rounding rules will apply:
 *
 * If the following decimal digits are less than 50, 500, 5000 ..., the previous one does not change.
 * If the following decimal digits are greater than 50, 500, 5000 ..., the previous one is incremented by one.
 * If the following decimal digits are equal to 50, 500, 5000 ..., the former is verified; if it is even, the former
 * does not change; if it is odd, the former is increased by one.
 *
 * Examples Edit
 * Rounding to 2 decimal digits, we must pay attention to the third and fourth decimal. Thus, according to the previous
 * rules: The number 12.6529 would be rounded up to 12.65 (here it is 12.65, since 29 is less than 50, so it does not
 * change) The number 12.86512 would be rounded to 12.87 (here it is 12.87, since 512 is greater than 500, then a unit
 * is increased) The number 12.744623 would be rounded to 12.74 (here it is 12.74, since 4623 is less than 5000, so it
 * does not change) The number 12.8752 would be rounded to 12.88 (here it is 12.88, since 52 is greater than 50, then a
 * unit is increased) The number 12.8150 would be rounded to 12.82 (here it is 12.82, since the next digits are equal
 * to 50 and the previous one is odd, in this case 1, then a unit is increased) The number 12.8050 would be rounded to
 * 12.80 (here it is 12.80, since the following digits are equal to 50 and the previous one is even, in this case 0,
 * then the previous one does not change)
 */
class AbntNbr5891
{
    /** Round number using brazilian normative ABNT NBR 5891
     *
     * @param $number
     *
     * @return mixed
     * @throws CommaException
     */
    public static function round($number)
    {
        $number = (string)$number;

        if (strpos($number, ',') !== false) {
            throw new CommaException();
        }

        $numberArr = explode('.', "" . $number);

        $intPart = $numberArr[0];
        $decimalsNumbers = isset($numberArr[1]) ? (string)$numberArr[1] : '00';

        $down = false;
        $up = false;

        if (strlen($decimalsNumbers) <= 2) {
            $decimalsNumbers = str_pad($decimalsNumbers, 2, '0', STR_PAD_RIGHT);
            return ((float)($intPart . '.' . $decimalsNumbers));
        }

        $decimalsStr = substr($decimalsNumbers, 0, 2);
        $restStr = substr($decimalsNumbers, 2);

        if (strlen($restStr) == 1) {
            $restStr = $restStr . '0';
        }

        $strlenRest = strlen($restStr);
        $finalRest = str_pad('5', $strlenRest, '0', STR_PAD_RIGHT);

        if ($restStr < $finalRest) {
            $down = true;
            $up = false;
        } else {
            if ($restStr > $finalRest) {
                $down = false;
                $up = true;
            } else {
                if ($restStr == $finalRest) {
                    if (((int)$decimalsStr[1]) % 2 == 1) {
                        $down = false;
                        $up = true;
                    } else {
                        $down = true;
                        $up = false;
                    }
                }
            }
        }

        $final = 0;
        if ($down) {
            $final = ((float)($intPart . '.' . $decimalsStr));
        } else {
            if ($up) {
                $decimals = $decimalsStr + 1;
                $sumInt = $decimals == 100 ? 1 : 0;
                if ($sumInt > 0) {
                    $decimals = '00';
                }

                if ($decimalsStr[0] == 0) {
                    $decimals = '0' . $decimals;
                }

                $final = ((float)(($intPart + $sumInt) . '.' . ($decimals)));
            }
        }

        return $final;
    }

    /**
     * @param      $value
     * @param int  $percent
     * @param bool $rounded
     *
     * @return float
     */
    public static function getPercentFromAmount($value, $percent = 100): float
    {
        return (float)($value / 100 * $percent);
    }
}
